File: frontend/src/types/preview.ts
export interface TrendPoint {
  date: string;
  value: number;
}

export interface PreviewGoal {
  entity_type: "goal";
  entity_id: string;
  title: string;
  status: string;
  progress: number;
  target_value: number;
  current_value: number;
  deadline: string | null;
}

export interface PreviewMetricType {
  entity_type: "metric_type";
  entity_id: string;
  title: string;
  unit: string;
  latest_value: number | null;
  latest_date: string | null;
  trend: TrendPoint[];
}

export interface PreviewExerciseType {
  entity_type: "exercise_type";
  entity_id: string;
  title: string;
  category: string;
  result_unit: string;
  pr_value: number | null;
  pr_date: string | null;
  recent_results: TrendPoint[];
}

export type EntityPreviewData = PreviewGoal | PreviewMetricType | PreviewExerciseType;

File: backend/app/schemas/preview.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class TrendPoint(BaseModel):
    date: date
    value: float


class PreviewGoal(BaseModel):
    entity_type: str = "goal"
    entity_id: str
    title: str
    status: str
    progress: float
    target_value: float
    current_value: float
    deadline: date | None


class PreviewMetricType(BaseModel):
    entity_type: str = "metric_type"
    entity_id: str
    title: str
    unit: str
    latest_value: float | None
    latest_date: date | None
    trend: list[TrendPoint]


class PreviewExerciseType(BaseModel):
    entity_type: str = "exercise_type"
    entity_id: str
    title: str
    category: str
    result_unit: str
    pr_value: float | None
    pr_date: date | None
    recent_results: list[TrendPoint]
